### PR TITLE
SLVS-962

### PR DESCRIPTION
**Pull request by tvsonar**

### CODE REVIEW

File: src/SonarLint/Helpers/SymbolicExecution/ProgramState.cs [hunk 1]
```diff
@@ -49,25 +50,90 @@ public class ProgramState : IEquatable<ProgramState>
...
+
+        public ProgramState PopValue()
+        {
+            SymbolicValue poppedValue;
+            return PopValue(out poppedValue);
+        }
+
+        public ProgramState PopValue(out SymbolicValue poppedValue)
+        {
+            return new ProgramState(
```
Comment (SonarTech) [inline src/SonarLint/Helpers/SymbolicExecution/ProgramState.cs]: MAJOR Consider a design that does not require that 'poppedValue' be an out parameter. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AAvoidOutParameters)

File: src/SonarLint.CSharp/Helpers/SymbolicExecution/ExplodedGraph.cs [hunk 2]
```diff
@@ -438,43 +467,21 @@ private void VisitInstruction(Node node)
...
-                        }
-                    }
+                    newProgramState = VisitSimpleAssignment((AssignmentExpressionSyntax)instruction, newProgramState);
                     break;
+
                 case SyntaxKind.OrAssignmentExpression:
+                    newProgramState = VisitBooleanBinaryOpAssignment(newProgramState, (AssignmentExpressionSyntax)instruction, (l, r) => new OrSymbolicValue(l, r));
+                    break;
                 case SyntaxKind.AndAssignmentExpression:
+                    newProgramState = VisitBooleanBinaryOpAssignment(newProgramState, (AssignmentExpressionSyntax)instruction, (l, r) => new AndSymbolicValue(l, r));
```
Comment (SonarTech) [inline src/SonarLint.CSharp/Helpers/SymbolicExecution/ExplodedGraph.cs]: MAJOR 'instruction', a variable, is cast to type 'AssignmentExpressionSyntax' multiple times in method 'ExplodedGraph.VisitInstruction(ExplodedGraph.Node)'. Cache the result of the 'as' operator or direct cast in order to eliminate the redundant castclass instruction. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3ADoNotCastUnnecessarily)

File: src/SonarLint.CSharp/Helpers/SymbolicExecution/ExplodedGraph.cs [hunk 3]
```diff
@@ -212,223 +209,255 @@ private void OnProgramPointVisitCountExceedLimit(ProgramPoint programPoint, Prog
...
+            return binaryBranchBlock.BranchingNode.IsKind(SyntaxKind.LogicalOrExpression)
+                ? newProgramState.PushValue(SymbolicValue.True)
+                : newProgramState;
         }
 
         #endregion
 
         private void VisitInstruction(Node node)
         {
             var instruction = node.ProgramPoint.Block.Instructions[node.ProgramPoint.Offset];
```
Comment (SonarTech) [inline src/SonarLint.CSharp/Helpers/SymbolicExecution/ExplodedGraph.cs]: MAJOR 'ExplodedGraph.VisitInstruction(ExplodedGraph.Node)' has a cyclomatic complexity of 65. Rewrite or refactor the method to reduce complexity to 25. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AAvoidExcessiveComplexity)

File: src/SonarLint.CSharp/Helpers/SymbolicExecution/ExplodedGraph.cs [hunk 4]
```diff
@@ -438,43 +467,21 @@ private void VisitInstruction(Node node)
...
-                            var constValue = SemanticModel.GetConstantValue(assignment.Right);
-
-                            newProgramState = GetNewProgramStateForAssignment(node.ProgramState, leftSymbol, rightSymbol, constValue);
-                        }
-                    }
+                    newProgramState = VisitSimpleAssignment((AssignmentExpressionSyntax)instruction, newProgramState);
                     break;
+
                 case SyntaxKind.OrAssignmentExpression:
+                    newProgramState = VisitBooleanBinaryOpAssignment(newProgramState, (AssignmentExpressionSyntax)instruction, (l, r) => new OrSymbolicValue(l, r));
```
Comment (SonarTech) [inline src/SonarLint.CSharp/Helpers/SymbolicExecution/ExplodedGraph.cs]: MAJOR 'instruction', a variable, is cast to type 'AssignmentExpressionSyntax' multiple times in method 'ExplodedGraph.VisitInstruction(ExplodedGraph.Node)'. Cache the result of the 'as' operator or direct cast in order to eliminate the redundant castclass instruction. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3ADoNotCastUnnecessarily)

File: src/SonarLint.CSharp/Helpers/SymbolicExecution/ExplodedGraph.cs [hunk 5]
```diff
@@ -484,59 +491,400 @@ private void VisitInstruction(Node node)
 
                 case SyntaxKind.LeftShiftAssignmentExpression:
                 case SyntaxKind.RightShiftAssignmentExpression:
-                    {
-                        var assignment = (AssignmentExpressionSyntax)instruction;
-                        var leftSymbol = SemanticModel.GetSymbolInfo(assignment.Left).Symbol;
-                        newProgramState = SetNewSymbolicValueIfLocal(newProgramState, leftSymbol);
-                    }
+                    newProgramState = VisitOpAssignment((AssignmentExpressionSyntax)instruction, newProgramState);
```
Comment (SonarTech) [inline src/SonarLint.CSharp/Helpers/SymbolicExecution/ExplodedGraph.cs]: MAJOR 'instruction', a variable, is cast to type 'AssignmentExpressionSyntax' multiple times in method 'ExplodedGraph.VisitInstruction(ExplodedGraph.Node)'. Cache the result of the 'as' operator or direct cast in order to eliminate the redundant castclass instruction. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3ADoNotCastUnnecessarily)

File: src/SonarLint/Helpers/SymbolicExecution/SymbolicValues/SymbolicValue.cs [hunk 6]
```diff
@@ -72,88 +74,100 @@ public override string ToString()
...
             }
 
             var objectConstraint = constraint as ObjectConstraint;
             if (objectConstraint != null)
             {
-                return TrySetConstraint(objectConstraint, oldConstraint, currentProgramState, out newProgramState);
+                return TrySetConstraint(objectConstraint, oldConstraint, currentProgramState);
             }
 
             throw new NotSupportedException($"Neither {nameof(BoolConstraint)}, nor {nameof(ObjectConstraint)}");
```
Comment (SonarTech) [inline src/SonarLint/Helpers/SymbolicExecution/SymbolicValues/SymbolicValue.cs]: MAJOR Correct the spelling of the unrecognized token 'BoolConstraint' in the literal '"BoolConstraint"'. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3ALiteralsShouldBeSpelledCorrectly)
MAJOR Correct the spelling of the unrecognized token 'ObjectConstraint' in the literal '"ObjectConstraint"'. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3ALiteralsShouldBeSpelledCorrectly)

File: src/SonarLint/Helpers/SymbolicExecution/SymbolicValues/SymbolicValue.cs [hunk 7]
```diff
@@ -72,88 +74,100 @@ public override string ToString()
...
-                    return false;
+                    return Enumerable.Empty<ProgramState>();
                 }
 
-                newProgramState = SetConstraint(objectConstraint, currentProgramState);
-                return true;
+                return new[] { SetConstraint(objectConstraint, currentProgramState) };
             }
 
             throw new NotSupportedException($"Neither {nameof(BoolConstraint)}, nor {nameof(ObjectConstraint)}");
```
Comment (SonarTech) [inline src/SonarLint/Helpers/SymbolicExecution/SymbolicValues/SymbolicValue.cs]: MAJOR Correct the spelling of the unrecognized token 'BoolConstraint' in the literal '"BoolConstraint"'. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3ALiteralsShouldBeSpelledCorrectly)
MAJOR Correct the spelling of the unrecognized token 'ObjectConstraint' in the literal '"ObjectConstraint"'. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3ALiteralsShouldBeSpelledCorrectly)

**Comment by SonarTech** [comment]
SonarQube analysis reported 13 issues
* MAJOR 12 major
* INFO 1 info

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:D1AE804F-AE78-4883-B3D6-9E3C4026DEF6: Because field 'BranchBlock.successors' is visible outside of its declaring type, change its accessibility to private and add a property, with the same accessibility as the field has currently, to provide access to it. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3ADoNotDeclareVisibleInstanceFields)
1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:D1AE804F-AE78-4883-B3D6-9E3C4026DEF6: Because field 'LiveVariableAnalysis.capturedVariables' is visible outside of its declaring type, change its accessibility to private and add a property, with the same accessibility as the field has currently, to provide access to it. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3ADoNotDeclareVisibleInstanceFields)
1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:D1AE804F-AE78-4883-B3D6-9E3C4026DEF6: Because field 'UnarySymbolicValue.operand' is visible outside of its declaring type, change its accessibility to private and add a property, with the same accessibility as the field has currently, to provide access to it. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3ADoNotDeclareVisibleInstanceFields)
1. INFO [ControlFlowGraphBuilder.cs#L649](https://github.com/SonarSource-VisualStudio/sonaranalyzer-csharp/blob/eb769e66c51e2cb7cc08049c6ef1923eedb02059/src/SonarLint.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs#L649): Complete the task associated to this "TODO" comment. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. INFO [ControlFlowGraphBuilder.cs#L649](https://github.com/SonarSource-VisualStudio/sonaranalyzer-csharp/blob/03269ef4d313b4cfcf6941fd20d2faafec44d9da/src/SonarLint.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs#L649): Complete the task associated to this "TODO" comment. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/SonarLint.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs [hunk 8]
```diff
@@ -1057,5 +1057,11 @@ private void BuildVariableDeclarator(VariableDeclaratorSyntax variableDeclarator
         #endregion
 
         #endregion
+
+        private ForInitializerBlock CreateForInitializerBlock(ForStatementSyntax forStatement, Block successor) =>
```
Comment (benzonico) [inline src/SonarLint.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs]: Why having those methods ? why not inline them ?

File: src/SonarLint/Helpers/SymbolicExecution/ProgramState.cs [hunk 9]
```diff
@@ -75,6 +75,24 @@ public ProgramState PushValue(SymbolicValue symbolicValue)
                 ExpressionStack.Push(symbolicValue));
         }
 
+        public ProgramState PushValues(IEnumerable<SymbolicValue> values)
+        {
+            if (values == null ||
```
Comment (benzonico) [inline src/SonarLint/Helpers/SymbolicExecution/ProgramState.cs]: This should not accept null IEnumerables

File: src/SonarLint/Helpers/SymbolicExecution/ProgramState.cs [hunk 10]
```diff
@@ -75,6 +75,24 @@ public ProgramState PushValue(SymbolicValue symbolicValue)
...
+                !values.Any())
+            {
+                return this;
+            }
+
+            var newStack = ImmutableStack.Create(
```
Comment (benzonico) [inline src/SonarLint/Helpers/SymbolicExecution/ProgramState.cs]: why two lines ?

File: src/SonarLint.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs (hunk 8 above)
Comment (tvsonar) [inline src/SonarLint.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs]: It fit into the previous pattern, but they are only used in one place, so makes sense to inline them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. INFO [ControlFlowGraphBuilder.cs#L659](https://github.com/SonarSource-VisualStudio/sonaranalyzer-csharp/blob/332cb5e75d99b2f7b69c7b2ea7de44e618a8d4a6/src/SonarLint.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs#L659): Complete the task associated to this "TODO" comment. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. INFO [ControlFlowGraphBuilder.cs#L659](https://github.com/SonarSource-VisualStudio/sonaranalyzer-csharp/blob/d8f08eafb9d8cbb12e30a76af68c105747af52f3/src/SonarLint.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs#L659): Complete the task associated to this "TODO" comment. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

**Pull request merged**